ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables/Utils.hs

```diff
@@ -0,0 +1,327 @@
+-- ledger state. Returns ledger tables.
+prependDiffs' ::
+     (Castable l l'', Castable l' l'', HasLedgerTables l, HasLedgerTables l')
+  => l DiffMK -> l' DiffMK -> LedgerTables l'' DiffMK
+prependDiffs' l1 l2 = ltliftA2 rawPrependDiffs (ltprj l1) (ltprj l2)
+
+-- | Like 'prependDiffs'', but puts the ledger tables inside the second ledger
+-- state.
+prependDiffs ::
+     (Castable l l', HasLedgerTables l, HasLedgerTables l')
+  => l DiffMK -> l' DiffMK -> l' DiffMK
+prependDiffs l1 l2 = over l2 $ prependDiffs' l1 l2
+
+--
+-- Apply diffs
+--
+
+rawApplyDiffs ::
+     Ord k
+  => ValuesMK k v -- ^ Values to which differences are applied
+  -> DiffMK   k v -- ^ Differences to apply
+  -> ValuesMK k v
+rawApplyDiffs (ValuesMK vals) (DiffMK diffs) = ValuesMK (Diff.applyDiff vals diffs)
+
+-- | Apply diffs from the second ledger state to the values of the first ledger
+-- state. Returns ledger tables.
+applyDiffs' ::
+     (Castable l l'', Castable l' l'', HasLedgerTables l, HasLedgerTables l')
+  => l ValuesMK -> l' DiffMK -> LedgerTables l'' ValuesMK
+applyDiffs' l1 l2 = ltliftA2 rawApplyDiffs (ltprj l1) (ltprj l2)
+
+-- | Like 'applyDiffs'', but puts the ledger tables inside the second ledger
+-- state.
+applyDiffs ::
+     (Castable l l', HasLedgerTables l, HasLedgerTables l')
+  => l ValuesMK -> l' DiffMK -> l' ValuesMK
+applyDiffs l1 l2 = over l2 $ applyDiffs' l1 l2
+
+rawApplyDiffForKeys ::
+     Ord k
+  => ValuesMK k v
+  -> KeysMK k v
+  -> DiffMK k v
+  -> ValuesMK k v
+rawApplyDiffForKeys (ValuesMK vals) (KeysMK keys) (DiffMK diffs) =
+  ValuesMK (Diff.applyDiffForKeys vals keys diffs)
+
+applyDiffForKeys' ::
+     (Castable l l'', Castable l l', HasLedgerTables l, HasLedgerTables l')
+  => l ValuesMK -> LedgerTables l KeysMK -> l' DiffMK -> LedgerTables l'' ValuesMK
+applyDiffForKeys' l1 l2 l3 = ltliftA3 rawApplyDiffForKeys (ltprj l1) (castLedgerTables l2) (ltprj l3)
+
+applyDiffForKeys ::
+     (Castable l l', HasLedgerTables l, HasLedgerTables l')
+  => l ValuesMK -> LedgerTables l KeysMK -> l' DiffMK -> l' ValuesMK
+applyDiffForKeys l1 l2 l3 = over l3 $ applyDiffForKeys' l1 l2 l3
+
+applyDiffForKeys'onTables ::
+     (Castable l l'', Castable l l', HasLedgerTables l, HasLedgerTables l')
+  => LedgerTables l ValuesMK -> LedgerTables l KeysMK -> l' DiffMK -> LedgerTables l'' ValuesMK
+applyDiffForKeys'onTables l1 l2 l3 = ltliftA3 rawApplyDiffForKeys (castLedgerTables l1) (castLedgerTables l2) (ltprj l3)
+
+applyDiffForKeysOnTables ::
+     (Castable l l', HasLedgerTables l, HasLedgerTables l')
+  => LedgerTables l ValuesMK -> LedgerTables l KeysMK -> l' DiffMK -> l' ValuesMK
+applyDiffForKeysOnTables l1 l2 l3 = over l3 $ applyDiffForKeys'onTables l1 l2 l3
+
+--
+-- Calculate differences
+--
+
+rawCalculateDifference ::
+     (Ord k, Eq v)
+  => ValuesMK   k v
+  -> ValuesMK   k v
+  -> TrackingMK k v
+rawCalculateDifference (ValuesMK before) (ValuesMK after) = TrackingMK after (Diff.diff before after)
+
+calculateAdditions ::
```

Comment:
Now that I'm looking at the HFC chunk, I see this is ultimate used to compute `ValuesMK -> DiffMK`, where they're all insertions. So maybe just define that combinator directly instead of factoring this weird function through `forgetTrackingValues`.